### PR TITLE
testing: unit testing with tap

### PR DESCRIPTION
**Pull request by TheAlphaNerd**
This pull request is primarily for discussion and code review at the moment.

 * Old tests are removed and replace with a very high level integration test suite written with tape.
    * This test suite is not feature complete with the original yet
 * new script ```npm run watch``` which will rerun the test suite on edit 
 * the majority of functions from citgm.js have been abstracted into their own files for testing purposes

TODO:

- [x] Add Documentation about testing stack
- [x] Parity with original test suite
    - Currently not including tests for HMAC
- [x] Unit tests written for individual parts of the system

### CODE REVIEW

**Comment by TheAlphaNerd** [comment]
@jasnell this is ready for review

**Comment by jasnell** [comment]
Did a brief scan, overall LGTM!

File: README.md [hunk 1]
```diff
@@ -176,13 +186,13 @@ to identify the custom script for you.
 
 ### Additional Notes:
 
-* You may experience some wonkiness on Windows as I have not fully
-  tested the tool on that platform.
+* You may experience some wonkiness on Windows as the tool hs not been fully
```
Comment (jasnell) [inline README.md]: s/hs/has

Comment (TheAlphaNerd) [inline README.md]: fixed

**Comment by jasnell** [comment]
@rvagg .. when you get a moment, would appreciate your review on this also

**Comment by TheAlphaNerd** [comment]
modified to work on v0.12.x (test suite passing on osx)

**Comment by rvagg** [comment]
This is great .. but I don't have time to review it unfortunately. Rubber stamp lgtm, that's the best I can do.

@TheAlphaNerd: the Jenkins team agreed to make you an admin on ci.nodejs.org for the purpose of messing with the smoke testing functionality. We have https://ci.nodejs.org/job/smoker/ but don't feel limited to just one job if you want to play with different ways of attacking the problem. I'm not aware of any objections from @nodejs/build to re-using the existing test machines for smoke testing, no need for a special set of servers. Please don't involve the slaves that have "release" in their name or labels, however. Ask in #node-build if you have any questions.

**Comment by TheAlphaNerd** [comment]
Closes #17

**Comment by TheAlphaNerd** [comment]
Enables #18

**Pull request merged**